packages/kbn-esql-validation-autocomplete/src/shared/resources_helpers.ts

```diff
@@ -10,22 +10,65 @@
 import type { ESQLAst } from '@kbn/esql-ast';
 import type { ESQLCallbacks } from './types';
 import type { ESQLRealField } from '../validation/types';
+import { enrichFieldsWithECSInfo } from '../autocomplete/utils/ecs_metadata_helper';
 
 export function buildQueryUntilPreviousCommand(ast: ESQLAst, queryString: string) {
   const prevCommand = ast[Math.max(ast.length - 2, 0)];
   return prevCommand ? queryString.substring(0, prevCommand.location.max + 1) : queryString;
 }
 
+// ECS Metadata is very static, so info only needs to be fetched once
+export class EcsMetadataCache {
+  private static instance: EcsMetadataCache;
+  private cacheEcsMetadata: Record<string, { type: string; source: string }> | undefined;
+
+  private constructor() {}
+
+  public static getInstance(): EcsMetadataCache {
+    if (!EcsMetadataCache.instance) {
+      EcsMetadataCache.instance = new EcsMetadataCache();
+    }
+    return EcsMetadataCache.instance;
+  }
+
+  public getMetadata(): Record<string, { type: string; source: string }> | undefined {
+    return this.cacheEcsMetadata;
+  }
+
+  public setMetadata(metadata: Record<string, { type: string; source: string }>) {
+    this.cacheEcsMetadata = metadata;
+  }
+}
+
 export function getFieldsByTypeHelper(queryText: string, resourceRetriever?: ESQLCallbacks) {
   const cacheFields = new Map<string, ESQLRealField>();
+
+  const getEcsMetadata = async () => {
+    const cache = EcsMetadataCache.getInstance();
+    if (!cache.getMetadata() && resourceRetriever?.getFieldsMetadata) {
+      // Fetch full list of ECS field
+      const client = await resourceRetriever?.getFieldsMetadata();
+      if (client?.find) {
+        const metadata = await client.find({
+          attributes: ['type'],
+        });
+        cache.setMetadata(metadata?.fields);
+      }
+    }
+    return cache.getMetadata();
+  };
```

Comment:
> [!TIP]
> I believe you can simplify this part a lot by removing the `EcsMetadataCache` and updating `getEcsMetadata`.
> The fieldsMetadataClient is already a singleton instance (each time you call the `.getClient` method from the service you get the same client) and this already create a cached result for a requests for the given parameters.
> I think removing the cache and having a simple helper function should be enough to achieve the same result.
```ts
const getEcsMetadata = async () => {
  if (!resourceRetriever?.getFieldsMetadata) {
    return {};
  }

  // Fetch full list of ECS field
  const client = await resourceRetriever.getFieldsMetadata();
  return client.find({ attributes: ['type'] });
};
```

Reply:
Thanks for hint that it's already cached. Updated here [`536e748` (#192646)](https://github.com/elastic/kibana/pull/192646/commits/536e7488228442167c22da5667edb91b6d60d144)